Mobile profile view

brianlovin opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- mobile

Implements mobile profile views for:
- user
- community
- channel

Video: https://www.dropbox.com/s/irlg7w8m0las5yl/mobile-profiles.mov?dl=0

Known bugs:
- ~~home tab content disappears if you navigate away and then back. scrolling triggers a re-render~~ (EDIT(@mxstbr): Can't replicate)
- [x] i'm not sure how i'll handle nested scrollviews on profile pages
- [x] i don't know how to get the infinite pagination to trigger on profile views that have an outer scrollview cc @mxstbr

brianlovin committed:
eslint cleanup

brianlovin committed:
Make participant heads tappable

brianlovin committed:
Add navigation flow type util for now

brianlovin committed:
Make thread byline tappable

brianlovin committed:
Make message bylines tappable

brianlovin committed:
Style and layout for user profile header

brianlovin committed:
Styling cleanup and better loading indicators

brianlovin committed:
Make the profile tab work, fix edge cases in profile view

brianlovin committed:
Remove console logs

brianlovin committed:
Support navigation logic in avatar component

brianlovin committed:
Add community profile view

brianlovin committed:
Cleanup and implement channel profile view

mxstbr commented on mobile/components/Avatar/index.js:
```diff
@@ -1,18 +1,32 @@
 
+    if (navigate) {
+      return (
+        <TouchableHighlight onPress={() => navigate()}>
```
I'm not sure it makes sense to bake this into the `Avatar` component, I'm sure there'll be cases where we don't want it to be clickable.

Instead, let's move this outside to the individual uses so it's clear where it's clickable and where it isn't, and keep this component focused.

mxstbr replied:
After looking through how you use the `Avatar` component, I now see that `navigate` is actually passed from the props and not injected from the `withNavigation` call further down below, so that's actually not so bad.

Let's remove the unnecessary `withNavigation` wrapper and rename the prop `onPress`. (why should you only have to navigate on press?! not very intuitive)

mxstbr commented on mobile/components/ChatInput/index.js:
```diff
@@ -33,7 +33,7 @@ class ChatInput extends React.Component<Props, State> {
           placeholder="Your message here..."
           multiline
-          autoFocus
+          autoFocus={false}
```
I'm pretty sure we want this to auto focus? No?

brianlovin replied:
No, I don't think so - it's super frustrating to open threads and have the keyboard come up and block anything like scrolling or tapping onto any links. Maybe we can build out a better heuristic for this in the future.

mxstbr replied:
SGTM

mxstbr commented on mobile/components/Loading/index.js:
```diff
@@ -0,0 +1,23 @@
+import * as React from 'react';
+import { ActivityIndicator } from 'react-native';
+import { Container } from './style';
+import theme from '../theme';
```
Instead of importing the theme here, use `withTheme` from styled-components which injects the dynamic theme from the context to the props:

```JS
class Loading extends React.Component {
  render() {
    const { theme } = this.props;
    return ()
  }
}

export default compose(withTheme)(Loading);
```

That way, if we decide to add a night mode or something it'll work out of the box.

mxstbr commented on mobile/components/Messages/Author.js:
```diff
@@ -17,13 +22,19 @@ const AuthorWrapper = styled.View`
-      </Text>
-    </AuthorWrapper>
+    <TouchableHighlight
+      onPress={() => navigation.navigate(`User`, { id: author.user.id })}
```
Same thing here as with the Avatar—I can already think of a case where we don't want this component clickable (message replies), so let's move this a level up out, of this component, and keep it focused on its actual use.

mxstbr reviewed (commented):
_Amazing_ work! 😍😍😍

Just a couple component structural nitpicks inline—I wanna make sure the mobile component library is as well designed as it can be so we could eventually turn it into the the official component set/library.

mxstbr committed:
Add <ConditionalWrap> component to mobile

shoutout to @kitze https://mobile.twitter.com/thekitze/status/923146119181471745

mxstbr committed:
navigate -> onPress

mxstbr committed:
Cleanup review comments

mxstbr commented:
> home tab content disappears if you navigate away and then back

This doesn't happen for me

> i'm not sure how i'll handle nested scrollviews on profile pages
> i don't know how to get the infinite pagination to trigger on profile views that have an outer scrollview cc @mxstbr

The way to do this is to have a single one (for the threads) and the `FlatList` that renders them should have a [`ListHeaderComponent`](https://facebook.github.io/react-native/docs/flatlist.html#listheadercomponent) that renders the header profile so that it scrolls out of view when you scroll down.

mxstbr committed:
Merge branch 'alpha' into mobile-profile-view

mxstbr committed:
Allow all FlatList props to be passed to <ThreadFeed />

@brianlovin this allows us to do:

```JS
<ThreadFeed
  ListHeaderComponent={ProfileHeader}
/>
```

which solves the need for two scroll views and makes infinite scrolling
work!

mxstbr committed:
Make infinite loading work on community mobile profile view

mxstbr committed:
Fix infinite scrolling on all profile views

mxstbr commented:
@brianlovin fixed the infinite scrolling with the solution from the last comment! Look at 088b389 how easy that was 😍

mxstbr commented:
Everything is resolved apart from the home view content disappearing, which I can't replicate at all, so I'll ship this since I need the `ConditionalWrap` component for #3088 and some follow-up work!

brianlovin commented:
Sweet! Thanks for merging, I'll dig back into mobile today :)

brianlovin commented:
> @brianlovin fixed the infinite scrolling with the solution from the last comment! Look at 088b389 how easy that was 😍

@mxstbr - one downside of this approach is that it breaks a lot of the great parts of the apollo cache. Because the profile header is now wrapped inside the infinite list, it will not render until the infinite list query finishes loading. This means that going to a user, community, or channel profile now has a fullscreen loader even though all the necessary information for the header is in-cache.

We should rethink this, or modify the way the infinite list works so that it can render subcomponents while the outer query fetches the thread list.

mxstbr commented:
@brianlovin good point, hadn't considered that. Will give it a think, my first intuition would be to expand the InfiniteList to also handle loading states itself
